src/analysis/souffle/operations.dl

```diff
@@ -35,7 +35,34 @@
 // Operation universe.
 .decl isOperation(operation: Operation)
 
-isPrincipal(owner), isAccessPath(result) :- isOperation([owner, _, result, _, _]).
+// OperandList universe.
+.decl isOperandList(operandList: OperandList)
+
+// Operator universe.
+.decl isOperator(operator: Operator)
+
+// Attribute list universe.
+.decl isAttributeList(attributeList: AttributeList)
+
+// Fill the appropriate universes with the pieces of the operation.
+isPrincipal(owner), isOperator(operator), isAccessPath(result),
+isOperandList(operandList), isAttributeList(attributeList) :-
+isOperation([owner, operator, result, operandList, attributeList]).
+
+// Declare mappings from an operation to its fields.
+.decl operationHasResult(operation: Operation, result: AccessPath)
+.decl operationHasOwner(operation: Operation, owner: Principal)
+.decl operationHasOperator(operation: Operation, operator: Operator)
+.decl operationHasOperandList(operation: Operation, operandList: OperandList)
+.decl operationHasAttributeList(operation: Operation, attributeList: AttributeList)
+
+// Populate field association relations.
+operationHasResult(operation, result), operationHasOwner(operation, owner),
+operationHasOperator(operation, operator), operationHasOperandList(operation, operands),
+operationHasAttributeList(operation, attributes) :-
+  isPrincipal(owner), isOperator(operator), isAccessPath(result), isOperandList(operands),
+  isAttributeList(attributes), isOperation(operation),
```

Comment:
Is this needed?

Reply:
Yes, this is a multi-head rule that deconstructs all of the fields of the `operation` to populate the various field-access relations.

Reply:
I realized your question might be "are all of these universe groundings needed?" The answer to that is also yes:  the line `operation = [owner, operator, result, operands, attributes]` is not deconstructing the `operation` the way you might expect a pattern matching operation to do so; it is instead literally saying that there exists an `operation` that is equal to the record `[owner, operator, result, operands, attributes]`. It has the *effect* of pattern matching, but is not handled in a nice way, so we have to ground the members like this.